### PR TITLE
Let Huggingface Properly Initialize Arguments, and Fix FSDP-LORA Checkpoint-Saves and Resumption

### PR DESCRIPTION
**Pull request by fabianlim**
@raghukiran1224 @Ssukriti @anhuong Im suggesting two fixes here

1. We observe failures with newer version of transformers because of a newly added `xla_fsdp_v2` flag here. The current strategy is to manually patch missing flags because `transformers.TrainingArguments.__post_init__` is not called. But this means one has to constantly update the manual patches if there are changes to HF code, which is not ideal.  Huggingface trainer will properly initialize everything based on  `TrainingArguments` (including [gradient checkpointing](https://github.com/huggingface/transformers/blob/2cc8cf6ce7ae0416561acbb639df4bbc5f409b6f/src/transformers/trainer.py#L1749-L1755). 
2. The `PeftSavingCallback` is not the ideal patch now that https://github.com/huggingface/transformers/pull/28297 has been merged, where now FSDP will properly save and resume adapter checkpoints. This will supercede the `PeftSavingCallback` strategy, as `PeftSavingCallback` will not even properly handle resumptions and different state dict saving strategies, but the already merged fix will. the `PEFTCallback` is removed in `trl` [here](https://github.com/huggingface/trl/pull/1110)

### CODE REVIEW

**Comment by fabianlim** [comment]
I noticed that any changes here is also releated to this file https://github.ibm.com/ai-foundation/sft-trainer-image/blob/main/launch_training.py.

File: tuning/sft_trainer.py
Comment (Ssukriti) [inline tuning/sft_trainer.py:122]: does single GPU still work after this change? without `train_args.fsdp_config = {"xla": False}` I remember single GPU training was failing

File: tuning/sft_trainer.py
Comment (Ssukriti) [inline tuning/sft_trainer.py:236]: does this flag `trainer.is_fsdp_enabled` basically do the same thing where its enabled only in a multi GPU env that has the desired HW and disabled for single gPU env?

**Review by Ssukriti — CHANGES_REQUESTED** [review]
1. left some questions and comments
2. DCO check is failing
3. Now that we have linted and formatted our files in main branch, the PR will need to be rebased on latest main . you need to update your fork to the latest main and then this branch with `git merge main`
sorry for the trouble, but any major changes upstream with linting and formatting is done now, so its a one-time pain :) 

Thank you for the contribution

File: tuning/sft_trainer.py
Comment (fabianlim) [reply tuning/sft_trainer.py:122]: Yes I have verified that i runs in both single and multi gpu settings. 

if I can point your attention to [`TrainingArguments.__post_init__`](https://github.com/huggingface/transformers/blob/45c065109074d60c587d3e562f16531d02a422f6/src/transformers/training_args.py#L1691-L1707), you will see that it already has logic to assemble all these `self.fsdp_config` flags appropriately. 

As [it has been coming up](https://github.com/foundation-model-stack/fms-hf-tuning/pull/79#discussion_r1513617678) that we should keep `transformers < 4.38.0`, the reason is because at `4.38` they introduced a new `xla_fsdp_v2 ` flag, and it did not suffice to only set the `xla` flag. Hence, we should leave this to the `__post_init__`  

My guess is why it did not run before was that before this changes, the `__post_init__` of `TrainingArguments` is not called, and perhaps the flag `fsdp_config["xla"]` is

File: tuning/sft_trainer.py
Comment (fabianlim) [reply tuning/sft_trainer.py:110]: hmm. i ran the `trainer` with `num_train_epochs=0`, with a `linear` schedule and it didnt throw error, and returned me the final output

```
{'train_runtime': 0.2356, 'train_samples_per_second': 0.0, 'train_steps_per_second': 0.0, 'train_tokens_per_second': 0.0, 'train_loss': 0.0, 'epoch': 0}
0it [00:00, ?it/s]
```

My thinking is such a basic check, if is necessary will already been done by the HF `Trainer`

File: tuning/sft_trainer.py
Comment (fabianlim) [reply tuning/sft_trainer.py:236]: this is the API correct way to check if FSDP is enabled. it is [initialized like this](https://github.com/huggingface/transformers/blob/45c065109074d60c587d3e562f16531d02a422f6/src/transformers/trainer.py#L4101) from the `trainer.accelerator.state`, which is the official way HF handles FSDP configs.

File: tuning/sft_trainer.py
Comment (fabianlim) [reply tuning/sft_trainer.py]: Update: i put back the checks.

**Comment by fabianlim** [comment]
@Ssukriti I have rebased the changes and linted.

File: tuning/sft_trainer.py
Comment (Ssukriti) [reply tuning/sft_trainer.py:137]: thank you!

**Review by Ssukriti — APPROVED** [review]
Thanks a lot @fabianlim for thoroughly testing your PR!

File: tuning/sft_trainer.py
Comment (Ssukriti) [reply tuning/sft_trainer.py:137]: actually post_init is always called as it is a derived class. I think the error I was referring to was because in original first commit we were oevrriding it with the wrong value . we were doing `
        train_args.fsdp_config = {}` or something like that, so I changed it to `"xla": False` . But you are right that if we dont do anything the _post_init should take care of it. So makes sense to remove it, thanks

**Comment by Ssukriti** [comment]
cant merge because of failing pylint, which seems to be a valid warning. Since this is your first commit Fabian, I have to manually start the workflow checks - pylint etc, thats why you didnt catch it earlier. Once this PR is merged and your fork is recognized as a contributor, going forward  the pylint checks with run automatically on your PRs and you wont have to wait for an admin to start checks :) 

to run pylint locally on your machine you can do `tox -e lint` as documented here 
https://github.com/foundation-model-stack/fms-hf-tuning/pull/84/files

Our contributing guides should be merged soon. 

I have approved the PR and will merge as soon as pylint checks pass

File: tuning/config/configs.py
Comment (fabianlim) [reply tuning/config/configs.py]: ah ic. yes sorry it was my misunderstanding. I assumed the reason for putting in the `fsdp_config` hardcode was that `__post_init__` is not called automatically. But I have verified that it indeed is.

**Comment by fabianlim** [comment]
@Ssukriti I have added one commit on top of your merge above #84, that removes the extranous `__post_init__`. I have also ran the pylint (which checks ok). Can you help to trigger the approving workflows?

**Review by Ssukriti — APPROVED** [review]
thank you!!!

**Pull request merged**
